salt/_modules/better_kubernetes.py

```diff
@@ -0,0 +1,148 @@
+    if not HAS_LIBS:
+        return False, 'Missing Python `kubernetes.client` dependency'
+
+    return __virtualname__
+
+
+def _extract_obj_and_details(manifest):
+    try:
+        details = __utils__['kubernetes.get_object_details'](manifest)
+        obj = __utils__['kubernetes.convert_manifest_into_object'](manifest)
+    except ValueError as exc:
+        raise CommandExecutionError('Invalid manifest: {!s}'.format(exc))
+
+    return obj, details
+
+
+def _extract_kwargs(obj, details):
+    kwargs = {'name': obj.metadata.name}
+
+    if details.scope == 'namespaced':
+        kwargs['namespace'] = obj.metadata.namespace
+
+    if hasattr(details, 'group'):
+        # This object represents a CustomResource
+        kwargs.update({
+            'group': details.group,
+            'version': details.version,
+            'plural': details.plural,
+        })
+
+    return kwargs
+
+
+def get_object(manifest, kubeconfig, context):
+    obj, details = _extract_obj_and_details(manifest)
+
+    kwargs = _extract_kwargs(obj, details)
+
+    try:
+        result = details.retrieve(**kwargs)
+    except Exception as exc:  # TODO: which Exception could be raised, and why?
+        raise CommandExecutionError(
+            'Failed to retrieve object: {!s}'.format(exc)
+        )
+
+    if hasattr(result, 'to_dict'):
+        # TODO: check if useful, probably for CRs, could be nice to clean
+        result = result.to_dict()
+
+    return result
+
+
+def delete_object(manifest, kubeconfig, context):
+    obj, details = _extract_obj_and_details(manifest)
+
+    kwargs = {'name': obj.metadata.name, 'body': k8s_client.V1DeleteOptions()}
+    if details.scope == 'namespaced':
+        kwargs['namespace'] = obj.metadata.namespace
+
+    if hasattr(details, 'group'):
+        # This manifest is for a CustomObject
+        kwargs.update({
+            'group': details.group,
+            'version': details.version,
+            'plural': details.plural,
+        })
+
+    try:
+        details.delete(**kwargs)
+    except Exception as exc:
+        raise CommandExecutionError(
+            'Failed to delete object: {!s}'.format(exc)
+        )
+
+    return None
+
+
+def replace_object(manifest, kubeconfig, context):
+    obj, details = _extract_obj_and_details(manifest)
+
+    # TODO: make sure we can use the source manifest or if we need to use
+    #       `obj.to_dict()`
+    kwargs = {'name': obj.metadata.name, 'body': manifest}
+    if details.scope == 'namespaced':
+        kwargs['namespace'] = obj.metadata.namespace
+
+    try:
+        result = details.replace(**kwargs)
+    except Exception as exc:
+        raise CommandExecutionError(
+            'Failed to replace object: {!s}'.format(exc)
+        )
+
+    if hasattr(result, 'to_dict'):
+        # TODO: check if useful, probably for CRs, could be nice to clean
+        result = result.to_dict()
+
+    return result
+
+
+def create_object(name, kubeconfig, context, manifest):
```

Comment:
There's a bit of duplication in these functions, could abstract them out (e.g., `create_object = _object_manipulation_function('create')`, etc).

Reply:
We could, but I'd rather have clear helpers that explain what's happening and let these methods define the specificities, instead of a bunch of `if method == 'create':` tests in the generic `_object_manipulation_function(method)`.

If you disagree, please feel free to detail why :)

Reply:
Changed my mind when the methods got more complex :sweat_smile: